snippets/dig.md

```diff
@@ -0,0 +1,31 @@
+### dig
+
+Return the value in a nested JSON object by the key.
+
+Given the key name (or target), it will map through each key in the object and look for object type.
+If value is an object, dig is called recusively until the first matching key/value is found.
+
+```
+const dig = (obj, target) => {
```

Comment:
This can be greatly shortened using or( `maybe || eventually` ), `Object.values())`, and `Array.prototype.find()`

```js
dig = (obj, target) => obj[target] || Object.values(obj).map( val => typeof val !== 'object' ? false : dig(val, target)).find(Boolean)
```

Reply:
Good catch on the falsy!

I haven't used reduce lately to get information that might not exist so I've forgotten that's a good '1 pass' method. TBH a loop here is actually a better idea but not quite ES6+ enough for our style rules, so I keep forgetting reduce is a good compromise... the problems of doing 'always exists' development and using maybes... XD

Reply:
Is the check for `undefined` on the recursion necessary?

Reply:
I just realised the `undefined` check beneath was pointless. Also I'm not sure about the top check either... Check the edit.

Reply:
@atomiks I moved the `if (typeof val === 'object')` to a filter. I think it makes the `reduce` easier to read. What do you think about this version? 

```
const dig = (obj, target) =>
  target in obj 
    ? obj[target] 
    : Object
        .values(obj)
        .filter(value => typeof value === 'object')
        .reduce((acc, val) => acc !== undefined ? acc : dig(val, target), undefined)
```

Reply:
@hszeto reduce is to prevent multiple loops over the same data. We are 'reducing' the whole array to a single value. Filtering before that is an extra loop and code that we can avoid, but does make it more readable. It has a worse case of O(2n) and a best case of O(n+1) with a 'not found' case of O(n). Why go for something we don't know how long it could take with especially huge datasets possible instead of something we know how long it will take( O(n) )?

Reply:
@skatcat31 I am 100% with you in terms of performance. It's never a good idea to add more loops. In my mind, I was thinking of ways to make the code easier to read and understand for users to learn in 30 seconds.... I didn't put performance in priority. My bad. :P Anyway, I will make another adjustment and commit.